src/persist-client/src/machine.rs

```diff
@@ -0,0 +1,400 @@
+                let desc = Description::new(
+                    Antichain::from(lower),
+                    Antichain::from(upper),
+                    Antichain::from(since),
+                );
+                (key.clone(), desc)
+            })
+            .collect();
+        State {
+            writers,
+            readers,
+            trace,
+        }
+    }
+
+    pub fn to_meta(&self) -> CollectionMeta {
+        CollectionMeta {
+            writers: self
+                .writers
+                .iter()
+                .map(|(id, cap)| {
+                    let upper = cap.upper.elements().iter().map(|x| T::encode(x)).collect();
+                    let meta = WriterMeta {
+                        seqno: cap.seqno,
+                        upper,
+                    };
+                    (id.clone(), meta)
+                })
+                .collect(),
+            readers: self
+                .readers
+                .iter()
+                .map(|(id, cap)| {
+                    let since = cap.since.elements().iter().map(|x| T::encode(x)).collect();
+                    let meta = ReaderMeta {
+                        seqno: cap.seqno,
+                        since,
+                    };
+                    (id.clone(), meta)
+                })
+                .collect(),
+            trace: self
+                .trace
+                .iter()
+                .map(|(key, desc)| {
+                    let desc = TraceBatchMeta {
+                        lower: desc.lower().iter().map(|x| T::encode(x)).collect(),
+                        upper: desc.upper().iter().map(|x| T::encode(x)).collect(),
+                        since: desc.since().iter().map(|x| T::encode(x)).collect(),
+                    };
+                    (key.clone(), desc)
+                })
+                .collect(),
+            key_codec: "WIP".into(),
+            val_codec: "WIP".into(),
+            ts_codec: "WIP".into(),
+            diff_codec: "WIP".into(),
+        }
+    }
+
+    pub fn since(&self) -> Antichain<T> {
+        //  WIP double check this initial and the meet_assign
+        //
+        // WIP what happens if all readers go away and then some come back?
+        let mut ret = Antichain::from_elem(T::minimum());
+        for (_, cap) in self.readers.iter() {
+            ret.meet_assign(&cap.since);
+        }
+        ret
+    }
+
+    pub fn upper(&self) -> Antichain<T> {
+        // WIP double check this initial and the join_assign
+        //
+        // WIP what happens if all writers go away and then some come back?
+        let mut ret = Antichain::from_elem(T::minimum());
+        for (_, cap) in self.writers.iter() {
+            ret.join_assign(&cap.upper);
+        }
+        ret
+    }
+
+    pub fn seqno_since(&self) -> SeqNo {
+        let mut ret = None;
+        // WIP what happens if all writers go away and then some come back?
+        for (_, cap) in self.readers.iter() {
```

Comment:
I think this should look at the seqnos of all of the readers and all of the writers?

Reply:
I was thinking we need to explicitly model it in the state, like we do since

Q for you: I haven't been able to come up with a reason that we need the writer seqnos in state. can you? the reader ones are used to GC unused blobs, but in this structure, writers never read

Reply:
Debuggability maybe? Can't come up with a reason immediately, will continue to think about it